### ISSUE TITLE
Any lint suggestion that changes the type of an expression can break code

### ISSUE THREAD

**Issue by y21** [rust-lang/rust-clippy#11467]
### Description

There are a lot of lints with `MachineApplicable` suggestions that change the type of the expression, which can result in compile errors.

I looked through a bunch of lints to check where this could happen *(not an exhaustive list)*:
- cloned_instead_of_copied
- filter_map_bool_then
- filter_map_identity
- flat_map_identity
- manual_filter_map
- map_flatten
- map_identity
- unnecessary_to_owned

The general idea is that the expression might be used in a context where a specific type is required.
```rs
let _: std::iter::Map<_, _> = [1].iter().map(|i| i);
```
This results in a warning "unnecessary map of the identity function" and clippy suggests removing the `map` call, which results in a type check error. This is a silly and made up example, but there's a good number of open `I-suggestion-causes-errors` issues that all happen because of type changes made by machine applicable suggestions, like here.

I recently fixed a FP that was basically this (and this is also why I'm filing this issue; this "FP" can happen with many other lints):
The lint `useless_conversion` looks for calls like `f(x.into_iter())` where `f` is some function that has an `impl IntoIterator` parameter. It's unnecessary to call `into_iter` at call site because you can already pass in anything that is `IntoIterator`.
However, if the `x.into_iter()` expression is part of a macro call and bound to an `expr` metavaraible, it can then be expanded twice (once to trigger the lint, and a second time where the exact type is required):
```rs
fn f(_: impl IntoIterator) {}

macro_rules! x {
  ($e:expr) => {
    f($e); // unnecessary `.into_iter()`

    let _: std::array::IntoIter<_, 2> = $e; // `.into_iter()` needed here
  }
}

x!([1, 2].into_iter());
```
This used to suggest removing `.into_iter()` at the macro callsite, but doing that breaks code inside of the macro expansion (specifically the second `$e` expansion).
This sort of thing can be used to break any lint with a `MachineApplicable` suggestion that changes the type of an expression, unless we explicitly check if an expression appeared as a macro call argument.

Now to the actual question/discussion: should these be considered bugs/actual issues, or are these cases where the user should just `#[allow]` the lint?

To be even more pedantic, macros are kind of problematic on their own, because someone could write a macro that only works for a very specific token stream, and *any* suggestion will then lead to compile errors, but I guess we have to draw a line somewhere and this seems like a non-issue.
